prow/spyglass/viewers/buildlog/buildlog_viewer.go

```diff
@@ -0,0 +1,273 @@
+	GetMoreRequests    map[string]LogViewerRequest
+	RawGetAllRequests  map[string]string
+	RawGetMoreRequests map[string]string
+}
+
+// ViewHandler creates a view for a build log (or multiple build logs)
+func ViewHandler(artifacts []viewers.Artifact, raw string) string {
+	buildLogsView := BuildLogsView{
+		LogViews:           []LogArtifactView{},
+		GetAllRequests:     make(map[string]LogViewerRequest),
+		GetMoreRequests:    make(map[string]LogViewerRequest),
+		RawGetAllRequests:  make(map[string]string),
+		RawGetMoreRequests: make(map[string]string),
+	}
+	viewReq := LogViewerRequest{
+		Requests: make(map[string]LogViewData),
+	}
+	if err := json.Unmarshal([]byte(raw), &viewReq); err != nil {
+		logrus.WithError(err).Error("Error unmarshaling log view request data.")
+	}
+
+	getAllReq := LogViewerRequest{
+		Requests: make(map[string]LogViewData),
+	}
+
+	getMoreReq := LogViewerRequest{
+		Requests: make(map[string]LogViewData),
+	}
+
+	// Read log artifacts and construct template structs
+	for _, a := range artifacts {
+		viewData, ok := viewReq.Requests[a.JobPath()]
+		if !ok {
+			viewData = LogViewData{
+				UseBytes:  false,
+				HeadBytes: 0,
+				TailLines: 0,
+				Operation: "more",
+			}
+		}
+
+		logView, logViewData := nextViewData(a, viewData)
+
+		buildLogsView.LogViews = append(buildLogsView.LogViews, logView)
+		getAllReq.Requests[a.JobPath()] = logViewData
+		getMoreReq.Requests[a.JobPath()] = logViewData
+
+	}
+
+	// Build individualized requests for stateless callbacks
+	for _, a := range artifacts {
+		aName := a.JobPath()
+
+		buildLogsView.GetAllRequests[aName] = getAllReq
+		reqAll := buildLogsView.GetAllRequests[aName]
+
+		reqAllData := reqAll.Requests[aName]
+		reqAllData.Operation = "all"
+		reqAll.Requests[aName] = reqAllData
+		raw, err := json.Marshal(reqAll)
+		if err != nil {
+			logrus.WithError(err).Error("Failed to marshal build log more lines request object.")
+		}
+		buildLogsView.RawGetAllRequests[aName] = string(raw)
+
+		buildLogsView.GetMoreRequests[aName] = getMoreReq
+		reqMore := buildLogsView.GetMoreRequests[aName]
+
+		reqMoreData := reqMore.Requests[aName]
+		reqMoreData.Operation = "more"
+		reqMore.Requests[aName] = reqMoreData
+		raw, err = json.Marshal(reqMore)
+		if err != nil {
+			logrus.WithError(err).Error("Failed to marshal build log more lines request object.")
+		}
+		buildLogsView.RawGetMoreRequests[aName] = string(raw)
+	}
+
+	return LogViewTemplate(buildLogsView)
+}
+
+// nextViewData constructs the new log artifact view and needed requests from the current
+// view data and artifact
+func nextViewData(artifact viewers.Artifact, currentViewData LogViewData) (LogArtifactView, LogViewData) {
+	var logLines []string
+	var err error
+	newLogViewData := LogViewData{
+		Operation: "",
+	}
+	newArtifactView := LogArtifactView{
+		ArtifactName: artifact.JobPath(),
+		ArtifactLink: artifact.CanonicalLink(),
+		ViewName:     name,
+	}
+	switch currentViewData.Operation {
+	case "all":
```

Comment:
would benefit from

```go
type ViewOperation string
```

?

Reply:
Without proper string enums what does this actually gain us?